contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+    if (api.assert(lp, 'no existing liquidity position')
+      && api.assert(api.BigNumber(lp.baseQuantity).minus(baseQuantity).gte(0)
+      && api.BigNumber(lp.quoteQuantity).minus(quoteQuantity).gte(0), 'not enough liquidity in position')) {
+      lp.baseQuantity = api.BigNumber(lp.baseQuantity).minus(baseQuantity);
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).minus(quoteQuantity);
+      if (api.BigNumber(lp.baseQuantity).eq(0) && api.BigNumber(lp.quoteQuantity).eq(0)) {
+        await api.db.remove('liquidityPosition', lp);
+      } else {
+        await api.db.update('liquidityPosition', lp);
+      }
+
+      await api.transferTokens(api.sender, baseSymbol, baseQuantity, 'user');
+      await api.transferTokens(api.sender, quoteSymbol, quoteQuantity, 'user');
+      updatePoolStats(pool, -baseQuantity, -quoteQuantity, false);
+      api.emit('removeLiquidity', { memo: `Remove ${baseSymbol} and ${quoteSymbol}` });
+    }
+  }
+};
+
+actions.swapTokensForExactTokens = async (payload) => {
+  const {
+    tokenPair,
+    tokenSymbol,
+    tokenOut,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(typeof (tokenSymbol) === 'string', 'invalid token')
+    || !api.assert(tokenOut && api.BigNumber(tokenOut).gt(0), 'insufficient tokenOut')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  if (tokenSymbol !== baseSymbol) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
+
+  const tokenInAdjusted = api.BigNumber(getAmountIn(tokenOut, liquidityIn, liquidityOut));
+  if (!tokenInAdjusted.isFinite()) return;
+
+  const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: symbolIn });
+  const senderBaseFunded = api.BigNumber(senderBase.balance).gte(tokenInAdjusted);
+  const tokenPairDelta = tokenSymbol === baseSymbol ? [api.BigNumber(tokenOut).times(-1), tokenInAdjusted] : [tokenInAdjusted, api.BigNumber(tokenOut).times(-1)];
+  if (!api.assert(senderBaseFunded, 'insufficient input balance')
+    || !validateSwap(pool, tokenPairDelta[0], tokenPairDelta[1])) return;
+
+  await api.executeSmartContract('tokens', 'transferToContract', { symbol: symbolIn, quantity: tokenInAdjusted.toFixed(pool.precision), to: 'marketpools' });
+  await api.transferTokens(api.sender, symbolOut, tokenOut, 'user');
```

Comment:
You should put in a sanity check that the transferToContract action succeeded, before transferring tokens to api.sender.

Reply:
Should we emit anything if the transferToContract action fails or just return?

Reply:
I think it makes sense to handle it with api.assert so an error is recorded if it fails.  But either way is fine really, as the tokens contract action itself should also generate an error if something goes wrong.  There are examples of doing it both ways in the nft contract, using a helper function to check that the logs generated by the tokens contract action are as expected.